### PR TITLE
IAM-1402: use traefik route to expose charm

### PR DESCRIPTION
**Pull request by shipperizer**
- **feat: introduce public-route relation**
- **fix!: drop public and admin ingresses, rewrite internal ingress to internal route**
- **build: update traefik route lib**

**MANDATORY**: **release to a new track**

----

example usage

```
shipperizer in ~/shipperizer/iam-bundle-integration/examples/tutorial on main ● λ http 10.152.183.183/.well-known/jwks.json Hostname:iam.internal --follow --verify=false
HTTP/1.1 200 OK
Cache-Control: private, no-cache, no-store, must-revalidate
Content-Length: 1583
Content-Type: application/json; charset=utf-8
Date: Mon, 14 Jul 2025 12:45:51 GMT
Vary: Origin, Origin

{
    "keys": [
        {
            "alg": "RS256",
            "e": "AQAB",
            "kid": "7167215f-e59d-4c01-bb76-fbe74d1d73a3",
            "kty": "RSA",
            "n": "5OEX0Qei8rahAwLF2RDmrlRQbsBKw1OOK3KdrA2MLrq-f2pG0Wl7kXo3UKFM_Fvl80IWqyGJR4FcOIfhxyw3Gny1SYLqVrpoYyVBaKbRbl7trBtldX-xIxq4u2rX3gT1WfAV59z56K1zrBcVHvW1ouO2R4eV8eLqJnSWaGsfHCZAwWDLWzLoDaInYgUBs9ntLq3lRd1KBWb9sgnipPOAYWqbaDwHlwS5_zQ3u4CsNtyKZSejNXgS5bJT8DDd5RSMFxTWXOL3ytqLNREKK8HBcgsowRgUgAWLbCStk6Bmj4Wq4uMGtrYNpyawpWUJllAnje8ZgzLZlFf5APtuBzFOK5j1LTbi-R5ZGf4p3-oywJKTplBMWAtduw1pWFyXqbrrLHbRO1VfInjpz1n_YxVAxWoGbRWmyg16T9enFHjbzqc2qAtgMVWCyaX7Y2sjdE6ePzWtkSeBU65PCOnj0TRXTDRkKqr96PEkYSWUqDAjkU275Sx4zz0z9FMQYk_uPyjCZBe63k7pMCwSlbUZjZ07uj5mTy2TBA9v4odryBmJOukSkcSTKvjS0iIV3a9q95_RiIBcSfzmVsdqc6qXr9rvivtmo6OPBaJvzi55-AYStIPsslW-zehpYTRv84ZwmzZckPUsvGI91bGcOsoPPsIVj_SeOf3VM9vlFLuUCnxEpJ0",
            "use": "sig"
        },
        {
            "alg": "RS256",
            "e": "AQAB",
            "kid": "226f5d2c-b9d1-47fb-abbe-1eecd3c3a6fb",
            "kty": "RSA",
            "n": "wdTBNzyO914mqCZlyL-LWYocg5f8Oyay-V9gXAcu0tCIvT1gMtFoJLi1VN_wOYyD_oDCo5JCTv7awzlb-n8UWa7cAARj_DKdsk9wAkTbNuIod8Asy98zmavVXGtJpajLE4NbH-jYgAiw3JC4F78A7W-NGa_z725ytoQsymhZDyzaKbRcCMezgZ1MIwyUum11h2Bj-bJoiRA_OEMFWcX5S1E6edLuNS5agdXltln_ygre6iO7k5TsvVmo8ukeW32S8S16h5QKGdftQL2RRYfVqQakAiUeMQA5Vp6xYu7L0vWe_pprJxAMVtIRx8UiTiyPPa_c8Z2NAK5qxiQeGXj_sNJCyCswkF5UmnZVG39bwMeN2cWvUihUvSYVGQaXLZ47LhspkE-Bd2p0tXWBM-ovdIU_E3cn8gXFB7eSIlx1CNzf11uCTlFXxZSxJU8Htf2ZYWtZWkNCkrVKXijrd-z6SExxWxo-_4RW5fGWCet_UmyreQ9QbveZPReONxT1ng5-dwAw5U54ZEYZM3eLGSd9vOAshpXlDltWMxg3uM92-BImXewVd6jLVQcTY3DK5VRGFkENfKaJ0AbOoV-eFkGF_h-6aaRVP6M7kYE6Kt582chieSXX30ko9GMzraCmPv9geMGVtL9rInYBbxXTlbm50nfMhi0sB-Oeroe-JUPGb-0",
            "use": "sig"
        }
    ]
}


shipperizer in ~/shipperizer/iam-bundle-integration/examples/tutorial on main ● λ http 10.152.183.183/.well-known/openid-configuration Hostname:iam.internal --follow --verify=false
HTTP/1.1 200 OK
Cache-Control: private, no-cache, no-store, must-revalidate
Content-Length: 1936
Content-Type: application/json; charset=utf-8
Date: Mon, 14 Jul 2025 12:46:09 GMT
Vary: Origin, Origin

{
    "authorization_endpoint": "https://iam.internal/oauth2/auth",
    "backchannel_logout_session_supported": true,
    "backchannel_logout_supported": true,
    "claims_parameter_supported": false,
    "claims_supported": [
        "sub"
    ],
    "code_challenge_methods_supported": [
        "plain",
        "S256"
    ],
    "credentials_endpoint_draft_00": "https://iam.internal/credentials",
    "credentials_supported_draft_00": [
        {
            "cryptographic_binding_methods_supported": [
                "jwk"
            ],
            "cryptographic_suites_supported": [
                "PS256",
                "RS256",
                "ES256",
                "PS384",
                "RS384",
                "ES384",
                "PS512",
                "RS512",
                "ES512",
                "EdDSA"
            ],
            "format": "jwt_vc_json",
            "types": [
                "VerifiableCredential",
                "UserInfoCredential"
            ]
        }
    ],
    "device_authorization_endpoint": "https://iam.internal/oauth2/device/auth",
    "end_session_endpoint": "https://iam.internal/oauth2/sessions/logout",
    "frontchannel_logout_session_supported": true,
    "frontchannel_logout_supported": true,
    "grant_types_supported": [
        "authorization_code",
        "implicit",
        "client_credentials",
        "refresh_token"
    ],
    "id_token_signed_response_alg": [
        "RS256"
    ],
    "id_token_signing_alg_values_supported": [
        "RS256"
    ],
    "issuer": "https://iam.internal",
    "jwks_uri": "https://iam.internal/.well-known/jwks.json",
    "request_object_signing_alg_values_supported": [
        "none",
        "RS256",
        "ES256"
    ],
    "request_parameter_supported": true,
    "request_uri_parameter_supported": true,
    "require_request_uri_registration": true,
    "response_modes_supported": [
        "query",
        "fragment"
    ],
    "response_types_supported": [
        "code",
        "code id_token",
        "id_token",
        "token id_token",
        "token",
        "token id_token code"
    ],
    "revocation_endpoint": "https://iam.internal/oauth2/revoke",
    "scopes_supported": [
        "offline_access",
        "offline",
        "openid",
        "profile",
        "email",
        "phone"
    ],
    "subject_types_supported": [
        "public"
    ],
    "token_endpoint": "https://iam.internal/oauth2/token",
    "token_endpoint_auth_methods_supported": [
        "client_secret_post",
        "client_secret_basic",
        "private_key_jwt",
        "none"
    ],
    "userinfo_endpoint": "https://iam.internal/userinfo",
    "userinfo_signed_response_alg": [
        "RS256"
    ],
    "userinfo_signing_alg_values_supported": [
        "none",
        "RS256"
    ]
}


```

### CODE REVIEW

File: charmcraft.yaml [hunk 1]
```diff
@@ -57,6 +57,11 @@ requires:
     description: |
       Ingress used for cross-cluster communication where network topology is more complex
       than just one k8s cluster
+  public-route:
```
[the remark this answers is not in the document]
Comment (shipperizer) [reply charmcraft.yaml:60]: i'm concerned about backwards compatibility in this case, is there a way to deprecate?

File: src/charm.py [hunk 2]
```diff
@@ -346,36 +339,48 @@ def _on_config_changed(self, event: ConfigChangedEvent) -> None:
         self._holistic_handler(event)
         self._on_oauth_integration_created(event)
 
-    def _on_public_ingress_ready(self, event: IngressPerAppReadyEvent) -> None:
+    def _on_internal_ingress_joined(self, event: RelationJoinedEvent) -> None:
         self.unit.status = MaintenanceStatus("Configuring resources")
-        self._holistic_handler(event)
-        self._on_oauth_integration_created(event)
-        self._on_hydra_endpoints_ready(event)
+        self.internal_ingress._relation = event.relation
```
Comment (nsklikas) [inline src/charm.py:344]: If I understand correctly, the only reason we need this is to cater to unit tests and (MAYBE) edge cases around deferred events. While not ideal, I am fine with keeping this as we don't know how long it would take the mesh team to fix this.

BUT let's add a comment clarifying this for future reference and I think you should try adding this to the other ingress handlers as well to avoid adding hacks to the tests (like `reset_mocks` and mocking `secured`)

**Review by nsklikas — APPROVED** [review]

**Review by nsklikas — APPROVED** [review]
